packages/plugin-workspace-tools/sources/commands/foreach.ts

```diff
@@ -126,14 +134,22 @@ export default class WorkspacesForeachCommand extends BaseCommand {
       ? project.topLevelWorkspace
       : cwdWorkspace!;
 
+    const rootCandidates = this.since
+      ? Array.from(await gitUtils.fetchChangedWorkspaces({ref: this.since, project}))
+      : [rootWorkspace, ...(this.from.length > 0 ? rootWorkspace.getRecursiveWorkspaceChildren() : [])];
+
     const fromPredicate = (workspace: Workspace) => micromatch.isMatch(structUtils.stringifyIdent(workspace.locator), this.from);
     const fromCandidates: Array<Workspace> = this.from.length > 0
-      ? [rootWorkspace, ...rootWorkspace.getRecursiveWorkspaceChildren()].filter(fromPredicate)
-      : [rootWorkspace];
-
-    const candidates = this.recursive
-      ? [...fromCandidates, ...fromCandidates.map(candidate => [...candidate.getRecursiveWorkspaceDependencies()]).flat()]
-      : [...fromCandidates, ...fromCandidates.map(candidate => [...candidate.getRecursiveWorkspaceChildren()]).flat()];
+      ? rootCandidates.filter(fromPredicate)
+      : rootCandidates;
+
+    const candidates = new Set([...fromCandidates, ...(await Promise.all(fromCandidates.map(async candidate => [...(
+      this.recursive
+        ? this.since
+          ? await candidate.getRecursiveWorkspaceDependents()
+          : candidate.getRecursiveWorkspaceDependencies()
+        : candidate.getRecursiveWorkspaceChildren()
+    )]))).flat()]);
```

Comment:
I'm not sure I follow this piece of code; can you rewrite it in a more imperative way, similar to the commit I made?

Reply:
I had a stab at it today, and following what was done in `list.ts` tried to re-write it in a similar way. I'm not sure its better though. It seems like slightly more duplication. I also struggled to come up with a new variable name that covered all 3 cases.

```
let nestedCandidates;
if (this.recursive) {
  if (this.since) {
    await project.restoreInstallState();
    nestedCandidates = fromCandidates.map(candidate => candidate.getRecursiveWorkspaceDependents());
  } else {
    nestedCandidates = fromCandidates.map(candidate => candidate.getRecursiveWorkspaceDependencies());
  }
} else {
  nestedCandidates = fromCandidates.map(candidate => candidate.getRecursiveWorkspaceChildren());
}

const candidates = new Set(fromCandidates);
for (const dependents of nestedCandidates) {
  for (const candidate of dependents) {
    candidates.add(candidate);
  }
}
```